database/service_db.go

```
@@ -88,80 +87,65 @@ func (d *LevelDBServiceDB) All() ([]*service.Service, error) {
 				continue
 			}
+			iter.Release()
 			return nil, err
 		}
 		services = append(services, s)
 	}
 	iter.Release()
-	if err := iter.Error(); err != nil {
-		return nil, err
-	}
-
-	return services, nil
+	return services, iter.Error()
 }
 
 // Delete deletes service from database.
-// TODO this should also find and delete alias when id is given.
 func (d *LevelDBServiceDB) Delete(idOrAlias string) error {
-	idOrAliasBytes := []byte(idOrAlias)
-	id, err := d.aliases.Get(idOrAliasBytes, nil)
-	if err != nil && err != leveldb.ErrNotFound {
+	id, alias, err := d.getIDAndAlias(idOrAlias)
+	if err != nil {
 		return err
 	}
-	if id != nil { // has alias, deleting it first.
-		idOrAliasBytes = id
-		if err := d.aliases.Delete(idOrAliasBytes, nil); err != nil {
+	// delete alias
+	if alias != "" {
+		if err := d.aliases.Delete([]byte(alias), nil); err != nil {
 			return err
 		}
 	}
-	return d.db.Delete(idOrAliasBytes, nil)
+	// delete service
+	return d.db.Delete([]byte(id), nil)
 }
 
 // Get retrives service from database.
 func (d *LevelDBServiceDB) Get(idOrAlias string) (*service.Service, error) {
-	id, err := d.aliases.Get([]byte(idOrAlias), nil)
-	if err != nil && err != leveldb.ErrNotFound {
+	id, _, err := d.getIDAndAlias(idOrAlias)
+	if err != nil {
 		return nil, err
 	}
-	if string(id) != "" {
-		idOrAlias = string(id)
-	}
-	b, err := d.db.Get([]byte(idOrAlias), nil)
+	b, err := d.db.Get([]byte(id), nil)
 	if err != nil {
-		if err == leveldb.ErrNotFound {
-			return nil, &ErrNotFound{ID: idOrAlias}
-		}
 		return nil, err
 	}
 	return d.unmarshal(idOrAlias, b)
 }
 
 // Save stores service in database.
 func (d *LevelDBServiceDB) Save(s *service.Service) error {
+	// check service
 	if s.ID == "" {
 		return errCannotSaveWithoutID
 	}
+	// encode service
 	b, err := d.marshal(s)
 	if err != nil {
 		return err
 	}
+	// save service
+	if err := d.db.Put([]byte(s.ID), b, nil); err != nil {
+		return err
+	}
+	// save alias if exist
```

this shouldn't be optional and should return an error if not exists